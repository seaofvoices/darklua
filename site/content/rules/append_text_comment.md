---
description: Append a comment at the start or end of a file
added_in: "0.12.0"
parameters:
  - name: text
    type: string
    description: The string to use inside the comment (required if `file` is not defined)
  - name: file
    type: string
    description: A path to a file to be used as the comment content (required if `text` is not defined)
  - name: location
    default: start
    type: '"start" or "end"'
    description: The location where to add the comment
examples:
  - rules: "[{ rule: 'append_text_comment', text: 'hello!' }]"
    content: print('Print from module')
  - rules: "[{ rule: 'append_text_comment', text: 'hello!', location: 'end' }]"
    content: print('Print from module')
---

Use this rule to automatically insert a comment at the start or end of a file. This rule can be useful if you want to insert your license in each file.

**Note:** make sure to avoid using the `remove_comments` rule _after_ this rule in the process sequence, otherwise you will be removing your brand new comment.
