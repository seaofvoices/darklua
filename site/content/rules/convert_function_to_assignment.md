---
description: Convert global function definitions to assignment statements
added_in: "unreleased"
parameters: []
examples:
  - content: |
      function foo(a, b)
          return a + b
      end
  - content: |
      function obj:method(value)
          return self.field + value
      end
---

Global function declarations will be transformed into assignment statements. This rule handles simple global functions, functions with field access (`module.function`), and method definitions (`object:method`).

When converting method syntax (`:method`), the rule automatically adds `self` as the first parameter to maintain the same behavior.

Note that, depending on your Lua runtime implementation, you may no longer be able to use reflection-like APIs (eg `debug.info`) to acquire the name of the function, or the function name may be missing from stack traces of `error` invocations.
