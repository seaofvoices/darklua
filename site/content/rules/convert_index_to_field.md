---
description: Convert index expression into field expressions
added_in: "0.7.0"
parameters: []
examples:
  - content: "return var['field']"
  - content: "return { ['field'] = true }"
---

When an index expression is using a static string (or an expression that can be statically evaluated into a string), this rule replaces it with a field expression. This rule also applies for table declarations: an entry that uses the bracket syntax (e.g. `{ ["key"] = value }`) will get converted into a field-like entry when possible.
