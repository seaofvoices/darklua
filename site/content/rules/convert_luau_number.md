---
description: Convert Luau numbers into Lua compatible numbers
added_in: "unreleased"
parameters: []
examples:
  - content: "print(0b1000_0001)"
  - content: "return 0xA000_FFFF"
---

This rule transforms Luau-specific number literals into Lua compatible number literals. It converts binary literals (prefixed with `0b` or `0B`) to their hexadecimal equivalents. It also removes underscores used as digit separators in numbers.
