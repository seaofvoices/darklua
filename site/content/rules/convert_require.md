---
description: Convert require calls from one environment to another
added_in: "0.10.0"
parameters:
  - name: current
    required: true
    type: require mode
    description: The require mode used in the input code
  - name: target
    required: true
    type: require mode
    description: The require mode used to generate the new require calls
examples: []
---

This rule is particularly useful if you are writing Lua code that needs to be portable to Roblox, as you can automatically convert requires by file path to Roblox instances.

Right now, the current and target require modes have certain restrictions:

- current: can only be the `path` require mode
- target: can only be the `roblox` require mode

## Configuration Overview

Here is an overview of the rule configuration format:

```json5
{
  rule: "convert_require",
  current: {
    name: "path",

    // optional (defaults to 'init')
    module_folder_name: "init",

    // optional
    sources: {
      pkg: "./Packages",
    },
  },
  target: {
    name: "roblox",

    // optional
    rojo_sourcemap: "./path-to/sourcemap.json",

    // optional (defaults to 'find_first_child')
    indexing_style: "find_first_child", // 'wait_for_child' or 'property'
  },
}
```

For more information about how to configure each of require mode, visit the [path require mode documentation](/docs/path-require-mode/) and the [roblox require mode documentation](/docs/roblox-require-mode/).
