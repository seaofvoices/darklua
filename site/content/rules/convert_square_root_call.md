---
description: "Convert calls to `math.sqrt(value)` into an exponent form (`value ^ 0.5`)"
added_in: "0.17.0"
parameters: []
examples:
  - content: "local result = math.sqrt(16)"
  - content: "local result = math.sqrt(x + y)"
  - content: "local result = math.sqrt(calculate_value())"
---

This rule converts calls to `math.sqrt(value)` into the equivalent expression `value ^ 0.5`. This transformation can be useful in some Lua runtime as a performance optimization.
