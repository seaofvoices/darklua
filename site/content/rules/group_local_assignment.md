---
description: Groups local assignments into a single statement
added_in: "0.3.2"
parameters: []
examples:
  - content: |
      local foo = 1
      local bar = 2
---

This rule will merge consecutive local assignments.

The rule will not merge assignments if one assignment temporally depends on the previous one, since it would break the code or change the behavior. The following code would not be changed:

```lua
local foo = 1
local bar = foo
```

Since functions can return multiple values, assignments that extract more than one value will not get merged.

```lua
local foo, bar = multiple_return_values()
local baz = 0
```
