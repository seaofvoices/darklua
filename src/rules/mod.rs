//! A module that contains the different rules that mutates a Lua block.

mod append_text_comment;
pub mod bundle;
mod call_parens;
mod compute_expression;
mod configuration_error;
mod convert_index_to_field;
mod convert_require;
mod empty_do;
mod filter_early_return;
mod group_local;
mod inject_value;
mod method_def;
mod no_local_function;
mod remove_assertions;
mod remove_call_match;
mod remove_comments;
mod remove_compound_assign;
mod remove_debug_profiling;
mod remove_interpolated_string;
mod remove_nil_declarations;
mod remove_spaces;
mod remove_types;
mod remove_unused_variable;
mod rename_variables;
mod replace_referenced_tokens;
pub(crate) mod require;
mod rule_property;
mod shift_token_line;
mod unused_if_branch;
mod unused_while;

pub use append_text_comment::*;
pub use call_parens::*;
pub use compute_expression::*;
pub use configuration_error::RuleConfigurationError;
pub use convert_index_to_field::*;
pub use convert_require::*;
pub use empty_do::*;
pub use filter_early_return::*;
pub use group_local::*;
pub use inject_value::*;
pub use method_def::*;
pub use no_local_function::*;
pub use remove_assertions::*;
pub use remove_comments::*;
pub use remove_compound_assign::*;
pub use remove_debug_profiling::*;
pub use remove_interpolated_string::*;
pub use remove_nil_declarations::*;
pub use remove_spaces::*;
pub use remove_types::*;
pub use remove_unused_variable::*;
pub use rename_variables::*;
pub(crate) use replace_referenced_tokens::*;
pub use rule_property::*;
pub(crate) use shift_token_line::*;
pub use unused_if_branch::*;
pub use unused_while::*;

use crate::nodes::Block;
use crate::Resources;

use serde::de::{self, MapAccess, Visitor};
use serde::ser::SerializeMap;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use std::collections::HashMap;
use std::fmt;
use std::path::{Path, PathBuf};
use std::str::FromStr;

#[derive(Debug, Clone)]
pub struct ContextBuilder<'a, 'resources, 'code> {
    path: PathBuf,
    resources: &'resources Resources,
    original_code: &'code str,
    blocks: HashMap<PathBuf, &'a Block>,
    project_location: Option<PathBuf>,
}

impl<'a, 'resources, 'code> ContextBuilder<'a, 'resources, 'code> {
    pub fn new(
        path: impl Into<PathBuf>,
        resources: &'resources Resources,
        original_code: &'code str,
    ) -> Self {
        Self {
            path: path.into(),
            resources,
            original_code,
            blocks: Default::default(),
            project_location: None,
        }
    }

    pub fn with_project_location(mut self, path: impl Into<PathBuf>) -> Self {
        self.project_location = Some(path.into());
        self
    }

    pub fn build(self) -> Context<'a, 'resources, 'code> {
        Context {
            path: self.path,
            resources: self.resources,
            original_code: self.original_code,
            blocks: self.blocks,
            project_location: self.project_location,
        }
    }

    pub fn insert_block<'block: 'a>(&mut self, path: impl Into<PathBuf>, block: &'block Block) {
        self.blocks.insert(path.into(), block);
    }
}

/// The intent of this struct is to hold data shared across all rules applied to a file.
#[derive(Debug, Clone)]
pub struct Context<'a, 'resources, 'code> {
    path: PathBuf,
    resources: &'resources Resources,
    original_code: &'code str,
    blocks: HashMap<PathBuf, &'a Block>,
    project_location: Option<PathBuf>,
}

impl<'a, 'resources, 'code> Context<'a, 'resources, 'code> {
    pub fn block(&self, path: impl AsRef<Path>) -> Option<&Block> {
        self.blocks.get(path.as_ref()).copied()
    }

    pub fn current_path(&self) -> &Path {
        self.path.as_ref()
    }

    fn resources(&self) -> &Resources {
        self.resources
    }

    fn original_code(&self) -> &str {
        self.original_code
    }

    fn project_location(&self) -> &Path {
        self.project_location.as_deref().unwrap_or_else(|| {
            let source = self.current_path();
            source.parent().unwrap_or_else(|| {
                log::warn!(
                    "unexpected file path `{}` (unable to extract parent path)",
                    source.display()
                );
                source
            })
        })
    }
}

pub type RuleProcessResult = Result<(), String>;

/// Defines an interface that will be used to mutate blocks and how to serialize and deserialize
/// the rule configuration.
pub trait Rule: RuleConfiguration + fmt::Debug {
    /// This method should mutate the given block to apply the rule
    fn process(&self, block: &mut Block, context: &Context) -> RuleProcessResult;

    /// Return the list of paths to Lua files that is necessary to apply this rule. This will load
    /// each AST block from these files into the context object.
    fn require_content(&self, _current_source: &Path, _current_block: &Block) -> Vec<PathBuf> {
        Vec::new()
    }
}

pub trait RuleConfiguration {
    /// The rule deserializer will construct the default rule and then send the properties through
    /// this method to modify the behavior of the rule.
    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError>;
    /// This method should return the unique name of the rule.
    fn get_name(&self) -> &'static str;
    /// For implementing the serialize trait on the Rule trait, this method should return all
    /// properties that differs from their default value.
    fn serialize_to_properties(&self) -> RuleProperties;
    /// Returns `true` if the rule has at least one property.
    fn has_properties(&self) -> bool {
        !self.serialize_to_properties().is_empty()
    }
}

pub trait FlawlessRule {
    fn flawless_process(&self, block: &mut Block, context: &Context);
}

impl<T: FlawlessRule + RuleConfiguration + fmt::Debug> Rule for T {
    fn process(&self, block: &mut Block, context: &Context) -> RuleProcessResult {
        self.flawless_process(block, context);
        Ok(())
    }
}

/// A function to get the default rule stack for darklua. All the rules here must preserve all the
/// functionalities of the original code after being applied. They must guarantee that the
/// processed block will work as much as the original one.
pub fn get_default_rules() -> Vec<Box<dyn Rule>> {
    vec![
        Box::<RemoveSpaces>::default(),
        Box::<RemoveComments>::default(),
        Box::<ComputeExpression>::default(),
        Box::<RemoveUnusedIfBranch>::default(),
        Box::<RemoveUnusedWhile>::default(),
        Box::<FilterAfterEarlyReturn>::default(),
        Box::<RemoveEmptyDo>::default(),
        Box::<RemoveUnusedVariable>::default(),
        Box::<RemoveMethodDefinition>::default(),
        Box::<ConvertIndexToField>::default(),
        Box::<RemoveNilDeclaration>::default(),
        Box::<RenameVariables>::default(),
        Box::<RemoveFunctionCallParens>::default(),
    ]
}

pub fn get_all_rule_names() -> Vec<&'static str> {
    vec![
        APPEND_TEXT_COMMENT_RULE_NAME,
        COMPUTE_EXPRESSIONS_RULE_NAME,
        CONVERT_INDEX_TO_FIELD_RULE_NAME,
        CONVERT_LOCAL_FUNCTION_TO_ASSIGN_RULE_NAME,
        CONVERT_REQUIRE_RULE_NAME,
        FILTER_AFTER_EARLY_RETURN_RULE_NAME,
        GROUP_LOCAL_ASSIGNMENT_RULE_NAME,
        INJECT_GLOBAL_VALUE_RULE_NAME,
        REMOVE_ASSERTIONS_RULE_NAME,
        REMOVE_COMMENTS_RULE_NAME,
        REMOVE_COMPOUND_ASSIGNMENT_RULE_NAME,
        REMOVE_DEBUG_PROFILING_RULE_NAME,
        REMOVE_EMPTY_DO_RULE_NAME,
        REMOVE_FUNCTION_CALL_PARENS_RULE_NAME,
        REMOVE_INTERPOLATED_STRING_RULE_NAME,
        REMOVE_METHOD_DEFINITION_RULE_NAME,
        REMOVE_NIL_DECLARATION_RULE_NAME,
        REMOVE_SPACES_RULE_NAME,
        REMOVE_TYPES_RULE_NAME,
        REMOVE_UNUSED_IF_BRANCH_RULE_NAME,
        REMOVE_UNUSED_VARIABLE_RULE_NAME,
        REMOVE_UNUSED_WHILE_RULE_NAME,
        RENAME_VARIABLES_RULE_NAME,
    ]
}

impl FromStr for Box<dyn Rule> {
    type Err = String;

    fn from_str(string: &str) -> Result<Self, Self::Err> {
        let rule: Box<dyn Rule> = match string {
            APPEND_TEXT_COMMENT_RULE_NAME => Box::<AppendTextComment>::default(),
            COMPUTE_EXPRESSIONS_RULE_NAME => Box::<ComputeExpression>::default(),
            CONVERT_INDEX_TO_FIELD_RULE_NAME => Box::<ConvertIndexToField>::default(),
            CONVERT_LOCAL_FUNCTION_TO_ASSIGN_RULE_NAME => {
                Box::<ConvertLocalFunctionToAssign>::default()
            }
            CONVERT_REQUIRE_RULE_NAME => Box::<ConvertRequire>::default(),
            FILTER_AFTER_EARLY_RETURN_RULE_NAME => Box::<FilterAfterEarlyReturn>::default(),
            GROUP_LOCAL_ASSIGNMENT_RULE_NAME => Box::<GroupLocalAssignment>::default(),
            INJECT_GLOBAL_VALUE_RULE_NAME => Box::<InjectGlobalValue>::default(),
            REMOVE_ASSERTIONS_RULE_NAME => Box::<RemoveAssertions>::default(),
            REMOVE_COMMENTS_RULE_NAME => Box::<RemoveComments>::default(),
            REMOVE_COMPOUND_ASSIGNMENT_RULE_NAME => Box::<RemoveCompoundAssignment>::default(),
            REMOVE_DEBUG_PROFILING_RULE_NAME => Box::<RemoveDebugProfiling>::default(),
            REMOVE_EMPTY_DO_RULE_NAME => Box::<RemoveEmptyDo>::default(),
            REMOVE_FUNCTION_CALL_PARENS_RULE_NAME => Box::<RemoveFunctionCallParens>::default(),
            REMOVE_INTERPOLATED_STRING_RULE_NAME => Box::<RemoveInterpolatedString>::default(),
            REMOVE_METHOD_DEFINITION_RULE_NAME => Box::<RemoveMethodDefinition>::default(),
            REMOVE_NIL_DECLARATION_RULE_NAME => Box::<RemoveNilDeclaration>::default(),
            REMOVE_SPACES_RULE_NAME => Box::<RemoveSpaces>::default(),
            REMOVE_TYPES_RULE_NAME => Box::<RemoveTypes>::default(),
            REMOVE_UNUSED_IF_BRANCH_RULE_NAME => Box::<RemoveUnusedIfBranch>::default(),
            REMOVE_UNUSED_VARIABLE_RULE_NAME => Box::<RemoveUnusedVariable>::default(),
            REMOVE_UNUSED_WHILE_RULE_NAME => Box::<RemoveUnusedWhile>::default(),
            RENAME_VARIABLES_RULE_NAME => Box::<RenameVariables>::default(),
            _ => return Err(format!("invalid rule name: {}", string)),
        };

        Ok(rule)
    }
}

impl Serialize for dyn Rule {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let properties = self.serialize_to_properties();
        let property_count = properties.len();
        let rule_name = self.get_name();

        if property_count == 0 {
            serializer.serialize_str(rule_name)
        } else {
            let mut map = serializer.serialize_map(Some(property_count + 1))?;

            map.serialize_entry("rule", rule_name)?;

            let mut ordered: Vec<(String, RulePropertyValue)> = properties.into_iter().collect();

            ordered.sort_by(|a, b| a.0.cmp(&b.0));

            for (key, value) in ordered {
                map.serialize_entry(&key, &value)?;
            }

            map.end()
        }
    }
}

impl<'de> Deserialize<'de> for Box<dyn Rule> {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Box<dyn Rule>, D::Error> {
        struct StringOrStruct;

        impl<'de> Visitor<'de> for StringOrStruct {
            type Value = Box<dyn Rule>;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("rule name or rule object")
            }

            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
            where
                E: de::Error,
            {
                let mut rule: Self::Value = FromStr::from_str(value).map_err(de::Error::custom)?;

                rule.configure(RuleProperties::new())
                    .map_err(de::Error::custom)?;

                Ok(rule)
            }

            fn visit_map<M>(self, mut map: M) -> Result<Self::Value, M::Error>
            where
                M: MapAccess<'de>,
            {
                let mut rule_name = None;
                let mut properties = HashMap::new();

                while let Some(key) = map.next_key::<String>()? {
                    match key.as_str() {
                        "rule" => {
                            if rule_name.is_none() {
                                rule_name.replace(map.next_value::<String>()?);
                            } else {
                                return Err(de::Error::duplicate_field("rule"));
                            }
                        }
                        property => {
                            let value = map.next_value::<RulePropertyValue>()?;

                            if properties.insert(property.to_owned(), value).is_some() {
                                return Err(de::Error::custom(format!(
                                    "duplicate field {} in rule object",
                                    property
                                )));
                            }
                        }
                    }
                }

                if let Some(rule_name) = rule_name {
                    let mut rule: Self::Value =
                        FromStr::from_str(&rule_name).map_err(de::Error::custom)?;

                    rule.configure(properties).map_err(de::Error::custom)?;

                    Ok(rule)
                } else {
                    Err(de::Error::missing_field("rule"))
                }
            }
        }

        deserializer.deserialize_any(StringOrStruct)
    }
}

fn verify_no_rule_properties(properties: &RuleProperties) -> Result<(), RuleConfigurationError> {
    if let Some((key, _value)) = properties.iter().next() {
        return Err(RuleConfigurationError::UnexpectedProperty(key.to_owned()));
    }
    Ok(())
}

fn verify_required_properties(
    properties: &RuleProperties,
    names: &[&str],
) -> Result<(), RuleConfigurationError> {
    for name in names.iter() {
        if !properties.contains_key(*name) {
            return Err(RuleConfigurationError::MissingProperty(name.to_string()));
        }
    }
    Ok(())
}

fn verify_required_any_properties(
    properties: &RuleProperties,
    names: &[&str],
) -> Result<(), RuleConfigurationError> {
    if names.iter().any(|name| properties.contains_key(*name)) {
        Ok(())
    } else {
        Err(RuleConfigurationError::MissingAnyProperty(
            names.iter().map(ToString::to_string).collect(),
        ))
    }
}

fn verify_property_collisions(
    properties: &RuleProperties,
    names: &[&str],
) -> Result<(), RuleConfigurationError> {
    let mut exists = false;
    for name in names.iter() {
        if properties.contains_key(*name) {
            if exists {
                return Err(RuleConfigurationError::PropertyCollision(
                    names.iter().map(ToString::to_string).collect(),
                ));
            } else {
                exists = true;
            }
        }
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;

    #[test]
    fn snapshot_default_rules() {
        let rules = get_default_rules();

        assert_json_snapshot!("default_rules", rules);
    }

    #[test]
    fn snapshot_all_rules() {
        let rule_names = get_all_rule_names();

        assert_json_snapshot!("all_rule_names", rule_names);
    }

    #[test]
    fn verify_no_rule_properties_is_ok_when_empty() {
        let empty_properties = RuleProperties::default();

        assert_eq!(verify_no_rule_properties(&empty_properties), Ok(()));
    }

    #[test]
    fn verify_no_rule_properties_is_unexpected_rule_err() {
        let mut properties = RuleProperties::default();
        let some_rule_name = "rule name";
        properties.insert(some_rule_name.to_owned(), RulePropertyValue::None);

        assert_eq!(
            verify_no_rule_properties(&properties),
            Err(RuleConfigurationError::UnexpectedProperty(
                some_rule_name.to_owned()
            ))
        );
    }

    #[test]
    fn get_all_rule_names_are_deserializable() {
        for name in get_all_rule_names() {
            let rule: Result<Box<dyn Rule>, _> = name.parse();
            assert!(rule.is_ok(), "unable to deserialize `{}`", name);
        }
    }

    #[test]
    fn get_all_rule_names_are_serializable() {
        for name in get_all_rule_names() {
            let rule: Box<dyn Rule> = name
                .parse()
                .unwrap_or_else(|_| panic!("unable to deserialize `{}`", name));
            assert!(json5::to_string(&rule).is_ok());
        }
    }
}
