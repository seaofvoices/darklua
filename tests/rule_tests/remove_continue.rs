use darklua_core::rules::{RemoveContinue, Rule};

test_rule_snapshot!(
    remove_continue,
    RemoveContinue::default(),
    numeric_for_continue_first_case(
        r#"
    for i = 1, 10 do
        if i == 1 then
            continue
        end
        print(i)
    end
    "#
    ),
    generic_for_continue_or_break(
        r#"
    for key, value in array do
        if skip(key) then
            continue
        elseif stop(key) then
            break
        end
        print(value)
    end
    "#
    ),
    for_inside_while(
        r#"
    while notDone() do
        if hasEnded() then
            break
        end
        for i = 1, 10 do
            if i == 1 then
                continue
            end
            print(i)
        end
    end
    "#
    ),
    repeat_inside_for(
        r#"
    for key, value in array do
        if type(key) == 'number' then
            repeat
                key -= 1
            until key == 0
        elseif stop(key) then
            break
        end
        print(value)
    end
    "#
    ),
    nested_for_continue_statements(
        r#"
    for i = 1, 10 do
        for j = 1, 10 do
            if j % 2 == 0 then
                continue
            end
            print(i, j)
        end
    end
    "#
    ),
    multiple_conditions_continue(
        r#"
    for i = 1, 10 do
        if i < 3 or i > 8 then
            continue
        end
        print(i)
    end
    "#
    ),
    function_with_continue(
        r#"
    for i = 1, 10 do
        local function shouldSkip(val)
            return val % 2 == 0
        end
        if shouldSkip(i) then
            continue
        end
        print(i)
    end
    "#
    ),
    sequential_loops_with_continue(
        r#"
    for i = 1, 5 do
        if i % 2 == 0 then
            continue
        end
        print(i)
    end

    for j = 6, 10 do
        if j % 3 == 0 then
            continue
        end
        print(j)
    end
    "#
    ),
    for_continue_with_comment(
        r#"
    for i = 1, 10 do
        if i % 2 == 0 then
            continue -- Skip even numbers
        end
        -- Print odd numbers
        print(i)
    end
    "#
    ),
    for_loop_with_only_continue(
        r#"
    for i = 1, 10 do
        continue
    end
    "#
    ),
    for_loop_continue_in_function_expression(
        r#"
    for i = 1, 10 do
        array[i] = function()
            for j = i, i + 10 do
                if j % 2 == 0 then
                    continue
                else
                    print(i)
                end
            end
        end
    end
    "#
    ),
    for_loop_continue_in_function_statement(
        r#"
    for i = 1, 10 do
        local element = array[i]
        function element.call()
            for j = i, i + 10 do
                if j % 2 == 0 then
                    continue
                else
                    print(i)
                end
            end
        end
    end
    "#
    ),
);

test_rule_without_effects!(
    RemoveContinue::default(),
    numeric_for_loop(
        r#"
    for i = 1, 10 do
        print(i)
    end
    "#
    ),
    generic_for_loop(
        r#"
    for key, value in array do
        print(key, value)
    end
    "#
    ),
    while_loop_without_continue(
        r#"
    local i = 0
    while i < 10 do
        i += 1
        print(i)
    end
    "#
    ),
    repeat_until_without_continue(
        r#"
    local i = 0
    repeat
        i += 1
        print(i)
    until i >= 10
    "#
    ),
    nested_for_no_continue(
        r#"
    for i = 1, 5 do
        for j = 1, 5 do
            print(i, j)
        end
    end
    "#
    ),
    empty_numeric_for(
        r#"
    for i = 1, 10 do
    end
    "#
    ),
    numeric_for_comments_only(
        r#"
    for i = 1, 10 do
        -- This is a comment
        -- Another comment
    end
    "#
    ),
    numeric_for_with_if_break(
        r#"
    for i = 1, 10 do
        if i % 2 == 0 then
            break
        end
    end
    "#
    ),
);

#[test]
fn deserialize_from_object_notation() {
    json5::from_str::<Box<dyn Rule>>(
        r#"{
        rule: 'remove_continue'
    }"#,
    )
    .unwrap();
}
