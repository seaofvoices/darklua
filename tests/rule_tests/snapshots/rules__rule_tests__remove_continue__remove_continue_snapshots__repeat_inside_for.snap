---
source: tests/rule_tests/remove_continue.rs
expression: lua_code
---
    for key, value in array do
        if type(key) == 'number' then
            repeat
                key -= 1
            until key == 0
        elseif stop(key) then
            break
        end
        print(value)
    end
